pkg/operator/boundsatokensignercontroller/controller.go

```diff
@@ -0,0 +1,296 @@
+			kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().HasSynced,
+		},
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "BoundSATokenSignerController"),
+	}
+
+	kubeInformersForNamespaces.InformersFor(operatorNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().AddEventHandler(ret.eventHandler())
+
+	return ret
+}
+
+func (c *BoundSATokenSignerController) sync() error {
+	signingSecret, err := c.secretClient.Secrets(operatorNamespace).Get(NextSigningKeySecretName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	needKeypair := errors.IsNotFound(err) || len(signingSecret.Data[PrivateKeyKey]) == 0 || len(signingSecret.Data[PublicKeyKey]) == 0
+	if needKeypair {
+		newSecret, err := newSigningSecret()
+		if err != nil {
+			return err
+		}
+		signingSecret, _, err = resourceapply.ApplySecret(c.secretClient, c.eventRecorder, newSecret)
+		if err != nil {
+			return err
+		}
+		// requeue for after we should have recovered
+		c.queue.AddAfter(workQueueKey, readyInterval+10*time.Second)
+	}
+
+	certConfigMap, err := c.configMapClient.ConfigMaps(targetNamespace).Get(CertConfigMapName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		certConfigMap = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: targetNamespace,
+				Name:      CertConfigMapName,
+			},
+			Data: map[string]string{},
+		}
+	}
+	currPublicKey := string(signingSecret.Data[PublicKeyKey])
+	hasThisKey := false
+	for _, value := range certConfigMap.Data {
+		if value == currPublicKey {
+			hasThisKey = true
+			break
+		}
+	}
+	if !hasThisKey {
+		// Increment until a unique name is found
+		nextKeyIndex := len(certConfigMap.Data) + 1
+		nextKeyKey := ""
+		for len(nextKeyKey) == 0 {
+			possibleKey := fmt.Sprintf("service-account-%03d.pub", nextKeyIndex)
+			_, ok := certConfigMap.Data[possibleKey]
+			if !ok {
+				nextKeyKey = possibleKey
+				break
+			}
+			nextKeyIndex += 1
+		}
+
+		certConfigMap.Data[nextKeyKey] = currPublicKey
+		certConfigMap, _, err = resourceapply.ApplyConfigMap(c.configMapClient, c.eventRecorder, certConfigMap)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Check if next-sa-signing-key has been around long enough to be promoted.
+	// Giving time for apiserver instances to pick up the change in public keys before
+	// changing the private key minimizes the potential for one or more apiservers to
+	// issue tokens signed by the new private key that apiservers without the
+	// corresponding public key are unable to validate.
```

Comment:
so we lack back-pressure again here, right? If rolling update is blocked for some reason, we might get into trouble.

Reply:
Is this still a problem now that actual state is considered rather than just waiting for a random amount of time?

Reply:
Will read the new code. Looking at the deployed state should be enough.